src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -83,7 +84,32 @@ object Content {
   def fromLeafFiles(files: Seq[FileStatus]): Content = Content(Directory.fromLeafFiles(files))
 }
 
-case class Directory(name: String, files: Seq[FileInfo] = Seq(), subDirs: Seq[Directory] = Seq())
+case class Directory(
+    name: String,
+    files: Seq[FileInfo] = Seq(),
+    subDirs: Seq[Directory] = Seq()) {
+  def merge(that: Directory): Directory = {
+    if (name.equals(that.name)) {
+      val allFiles = files ++ that.files
+      val subDirMap = subDirs.map(dir => dir.name -> dir).toMap
+      val thatSubDirMap = that.subDirs.map(dir => dir.name -> dir).toMap
+      val subDir: Seq[Directory] = (subDirMap.keySet ++ thatSubDirMap.keySet).toSeq.map {
+        dirName =>
+          if (subDirMap.contains(dirName) && thatSubDirMap.contains(dirName)) {
+            // If both directories contain a subDir with same name, merge subDirs respectively.
+            subDirMap(dirName).merge(thatSubDirMap(dirName))
+          } else {
+            // Pick the subDir from whoever contains it.
+            subDirMap.getOrElse(dirName, thatSubDirMap(dirName))
+          }
+      }
+
+      Directory(name, allFiles, subDirs = subDir)
+    } else {
+      throw HyperspaceException("Directory names should match for merging Directories.")
```

Comment:
Can you improve this error message?

Reply:
thanks @rapoth , I couldn't quite get the point. Could you give a suggestion? I think it's descriptive. Here's what I mean "If two directories to be merged have different 'name' fields, throw exception".

Reply:
So think from a user standpoint. When this error pops up, it is not clear what directory is being referred to in this error. Also, what is a user supposed to do once they hit into this error? etc.

Reply:
thanks @rapoth , i updated the error message to display the names of directories being merged. I also added documentation to this api. Please let me know if this works